lib/streamlit/runtime/state/query_params.py

```diff
@@ -55,12 +56,23 @@ def __setitem__(self, key: str, value: Union[str, Iterable[str]]) -> None:
             raise StreamlitAPIException(
                 f"You cannot set a query params key `{key}` to a dictionary."
             )
+        # Avoid circular imports
+        from streamlit.commands.experimental_query_params import EMBED_QUERY_PARAMS_KEYS
 
         # Type checking users should handle the string serialization themselves
         # We will accept any type for the list and serialize to str just in case
         if isinstance(value, Iterable) and not isinstance(value, str):
+            for item in value:
+                if item in EMBED_QUERY_PARAMS_KEYS:
+                    raise StreamlitAPIException(
+                        "Query param embed and embed_options (case-insensitive) cannot be set programmatically."
+                    )
             self._query_params[key] = [str(item) for item in value]
         else:
+            if key in EMBED_QUERY_PARAMS_KEYS:
```

Comment:
need this explicit call because we are not calling this anymore:
```
msg.page_info_changed.query_string = _ensure_no_embed_params(
            self._query_params, ctx.query_string
)
```